### ISSUE TITLE
Cannot return null for non-nullable field Link.url for class based mutation

### ISSUE THREAD

**Issue by rainkinz** [rmosolgo/graphql-ruby#1647]
Hi,

I've been trying out the class based API following the example here: http://graphql-ruby.org/mutations/mutation_classes.html

I've defined my type like this:
```

**Comment by rmosolgo** [comment]
How odd!! Can you add `link.url.inspect` to your `logger.warn` message, just to make double-sure that the created link actually has a url?

**Comment by rainkinz** [comment]
I added this:
```
and get this:

```
Saved link id: '49' url: 'woot'
```

**Comment by rainkinz** [comment]
Oops, I'm sorry you asked for inspect, this is the output:
```
Saved link: '#<Link id: 50, url: "woot", description: "too", created_at: "2018-07-04 14:08:25", updated_at: "2018-07-04 14:08:25">'
```

**Comment by rainkinz** [comment]
As an aside, I'm pretty new to graphql, but to be honest, given my definition of the mutation above I would have expected the query to look like this:

```
mutation createLink {
  createLink(input:{url:"woot", description:"too"})
  {
      link{
        url
      }
  }
}
```
But that gives me the following error:

```
{
  "errors": [
    {
      "message": "Field 'link' doesn't exist on type 'Link'",
      "locations": [
        {
          "line": 45,
          "column": 7
        }
      ],
      "fields": [
        "mutation createLink",
        "createLink",
        "link"
      ]
    }
  ]
}
```

**Comment by rmosolgo** [comment]
Yeah, that's a great point! It _should_ have a `link { ... }` field ... why _doesn't_ it!? Can you share the output of `puts MySchema.to_definition`? (At least `type Mutation` and any of the `CreateLink*` types?) 

I wonder where it's getting confused!!

**Comment by rmosolgo** [comment]
Oh, also, could you share your base mutation class?

**Comment by rainkinz** [comment]
Of course! Thanks for for looking at this. The output of `puts GraphqlTutorialSchema.to_definition` is:

```
>> puts GraphqlTutorialSchema.to_definition
# Autogenerated input type of CreateLink
input CreateLinkInput {
  # A unique identifier for the client performing the mutation.
  clientMutationId: String
  description: String!
  url: String!
}

type Link {
  description: String
  id: ID!
  url: String!
}

type Mutation {
  createLink(input: CreateLinkInput!): Link
}

type Query {
  allLinks: [Link!]!

  # Find a link by id
  link(id: ID!): Link
}
=> nil
```

and the base mutation class looks like this:

```
class Mutations::BaseMutation < GraphQL::Schema::RelayClassicMutation
end
```

Note, this can be found here: https://github.com/rainkinz/ruby-graphql-tutorial/blob/master/app/graphql/mutations/base_mutation.rb  if you need a complete example.

I've also tried subclassing `GraphQL::Schema::Mutation`.

Thanks again!

**Comment by heitorado** [comment]
Hi @vfonic, I'm also getting this error for the same reason as yours. Even though this behavior is intended, did you manage to find a solution so that the error shown is something that makes more sense such as "Not autorized", or even returning just a 401 error code?

**Comment by bitkus** [comment]
> I got this error because `authorized?` method in my mutation was returning false which was causing fields to be nil. It would help if the error message would somehow point in the right direction.

Isn't this the point of authorized? According to the documentation, it should either throw or return a boolean indicating whether the subscription is authorized or not. How can you work around this then?

**Comment by vfonic** [comment]
It is the point, you're right! I'd prefer the error to be thrown rather than "failing silently". Why?

Because, just like how I would prefer the error to be thrown when I do:

```rb
user.asmin? # notice the typo: admin being misspelled as asmin
```

I don't want this line to fail silently and then I have to scratch my head why admins cannot perform some action. I prefer the application code to raise an error when there's an application error, an exception to the expected behavior.

Similarly above, user that doesn't have access to a particular field should never query for that field and expect to get any acceptable response. I'd return 500 server error in such cases and log the error to error tracking app.

If you query for a field that doesn't exist in the schema, you'll get an error. Then why wouldn't this return similar error?
"The field 'admin' does not exist on 'User' type or you don't have access to it"

This returning of `nil` for existing field vs returning error for non-existing field makes it possible for a hacker to guess my GraphQL schema, which could help him perform a hack in the future. (Maybe through CSRF, submitting as some other user?)
It's like when someone is trying to log in with wrong username/password combination. You don't write: wrong password (but email is correct), but you write "wrong email/password combination".

This is how my `authorized?` module looks like right now, so that I can avoid returning `nil`:

```rb
# frozen_string_literal: true

module Mutations::Authorization
  module UserIsLoggedIn
    def authorized?(**_input)
      super

      authorized = current_user.present?
      raise GraphQL::ExecutionError, "You must be logged in to run #{self.class.graphql_name} mutation" unless authorized

      authorized
    end
  end
end
```

It would be great if this was a part of the library (maybe a config option?).

/cc @rmosolgo

**Comment by rmosolgo** [comment]
>  "The field 'admin' does not exist on 'User' type or you don't have access to it"

It sounds like you're looking for "visibility": https://graphql-ruby.org/authorization/visibility.html . It allows you to conditionally hide parts of the schema from certain clients. It's different from authorization because visibility is applied during _validation_, not during query execution. 

As for raising an error, yes, I agree there are times when an error is the right thing. An approach like you described is also documented here: 

https://graphql-ruby.org/authorization/authorization.html#handling-unauthorized-objects

**Comment by vfonic** [comment]
That's amazing (https://graphql-ruby.org/authorization/visibility.html) and amazing (https://graphql-ruby.org/authorization/authorization.html#handling-unauthorized-objects)!

I missed this back in 2019, maybe it was added later or I just didn't see it while glancing through the documentation.

Thank you very much! This will simplify some of my codebase! 🚀
